frontend/catalyst/pennylane_extensions.py

```diff
@@ -2250,3 +2252,207 @@ def _callee():
+
+        # Prepare args_flat to run 'fn' one time and get the output-shape
+        fn_args_flat = args_flat.copy()
+        for loc in batch_loc:
+            fn_args_flat[loc] = args_flat[loc][0]
+
+        fn_args = tree_unflatten(args_tree, fn_args_flat)
+
+        # Run 'fn' one time to get output-shape
+        init_result = fn(*fn_args, **kwargs)
+
+        # Return 'init_result' if none is requested
+        if not batch_size:
+            return init_result
+
+        init_result_flat, init_result_tree = tree_flatten(init_result)
+
+        out_axes_flat = _vmap_tree_flatten(out_axes)
+        out_pos = _get_batch_loc(out_axes_flat)
+        out_pos = out_pos[0]
+
+        num_axes_out = len(init_result_flat)
+
+        # Store batched results of all leaves
+        # in the flatten format with respect to the 'init_result' shape
+        batched_result_list = []
+        for j in range(num_axes_out):
+            out_shape = (
+                (batch_size,)
+                if not init_result_flat[j].shape
+                else (batch_size, *init_result_flat[j].shape)
+            )
+            batched_result_list.append(jnp.zeros(shape=out_shape, dtype=init_result_flat[j].dtype))
+            batched_result_list[j] = batched_result_list[j].at[0].set(init_result_flat[j])
+
+        if isinstance(out_axes, tuple) and len(out_axes) != len(out_shape):
+            raise ValueError(
+                "Invalid 'out_axes'; it can be an int or a tuple of PyTrees corresponding to the "
+                f"arguments, but got out_axes={out_axes} and out_shape={out_shape}"
+            )
+
+        # TODO: replace for_loop with parallelfor_loop
+        # Apply mapping batched_args[1:] ---> fn(args)'
+        @for_loop(1, batch_size, 1)
+        def loop_fn(i, batched_result_list):
+            fn_args_flat = args_flat
+            for loc in batch_loc:
+                fn_args_flat[loc] = args_flat[loc][i]
+
+            fn_args = tree_unflatten(args_tree, fn_args_flat)
+            res = fn(*fn_args, **kwargs)
+
+            res_flat, _ = tree_flatten(res)
+
+            # Update the list of results
+            for j in range(num_axes_out):
+                batched_result_list[j] = batched_result_list[j].at[i].set(res_flat[j])
+
+            # Support out_axes on dim > 0
+            if ax := out_axes_flat[out_pos]:
+                up_axes = [*range(batched_result_list[out_pos].ndim)]
+                up_axes[ax], up_axes[0] = up_axes[0], up_axes[ax]
+                batched_result_list[out_pos] = jnp.transpose(batched_result_list[out_pos], up_axes)
+
+            return batched_result_list
+
+        batched_result_list = loop_fn(batched_result_list)
+
+        # Unflatten batched_result before return
+        return tree_unflatten(init_result_tree, batched_result_list)
+
+    return batched_fn
+
+
+def _vmap_tree_flatten(args):
+    """Custom tree_flatten for vmap to keep 'None' elements into the flatten list."""
```

Comment:
I don't fully follow this function, why is it needed?

Reply:
`tree_flatten` doesn't include `None` values as leaves. This function extends the support to also include `None` values in the list of flatten data for both `in_axes` and `out_axes`. I use this function to create a 1-to-1 map between `args` and `in_axes`, and between `return_shape` and `out_axes`. I'm updating the docstrings of all these methods providing more info and examples.